File: Sources/YBarKit/Items/Layout.swift
import CoreGraphics

/// Measured content of one item, produced by the text measurer.
public struct MeasuredContent {
    public var iconSize: CGSize
    public var labelSize: CGSize

    public init(iconSize: CGSize = .zero, labelSize: CGSize = .zero) {
        self.iconSize = iconSize
        self.labelSize = labelSize
    }
}

/// Pure layout of items along a horizontal bar — sketchybar's proven five-cursor
/// algorithm: left→, ←right, a centered center block (pre-pass sum), and
/// notch-anchored centerLeft/centerRight cursors flowing away from the dead zone.
///
/// All geometry is in bar-local points with a top-left origin. Layout writes
/// `item.frame` (the full interactive extent, bar height tall) and returns the
/// content boxes used by the renderer.
@MainActor
public enum Layout {
    public struct Result {
        /// Content box per item id: the area covered by background/text (without outer paddings).
        public var contentBoxes: [Int: CGRect] = [:]
    }

    /// Length of the item content: icon + label with their paddings (or fixed width).
    public static func contentLength(item: Item, measured: MeasuredContent) -> CGFloat {
        if item.customWidth >= 0 { return CGFloat(item.customWidth) }
        return naturalLength(item: item, measured: measured)
    }

    /// Content length ignoring any fixed-width override — the value `width=dynamic`
    /// animations resolve the -1 sentinel to.
    public static func naturalLength(item: Item, measured: MeasuredContent) -> CGFloat {
        // sketchybar parity: a drawing part contributes its paddings even with
        // an empty string (text_get_length gates on drawing alone).
        var length: CGFloat = 0
        if item.icon.drawing {
            length += partAdvance(item.icon, inkWidth: measured.iconSize.width)
        }
        length += item.contentWidth
        // A gauge item's label renders inside the dial, not beside it — it
        // must not widen the item (it skews align-slack centering).
        if item.label.drawing, item.gauge == nil {
            length += partAdvance(item.label, inkWidth: measured.labelSize.width)
        }
        return length
    }

    /// Total horizontal advance of a text part. A fixed customWidth REPLACES
    /// ink + paddings entirely (sketchybar's text_get_length override: paddings
    /// live inside the slot), so width=0 collapses to exactly zero.
    public static func partAdvance(_ part: TextPart, inkWidth: CGFloat) -> CGFloat {
        if part.customWidth >= 0 { return CGFloat(part.customWidth) }
        return CGFloat(part.paddingLeft) + inkWidth + CGFloat(part.paddingRight)
    }

    @discardableResult
    public static func perform(
        items: [Item],
        barSize: CGSize,
        settings: BarSettings,
        measure: (Item) -> MeasuredContent
    ) -> Result {
        var result = Result()
        let width = barSize.width
        let notchHalf = CGFloat(settings.notchWidth) / 2

        func place(_ item: Item, x: CGFloat, measured: MeasuredContent) {
            let length = contentLength(item: item, measured: measured)
            let box = CGRect(x: x, y: 0, width: length, height: barSize.height)
            item.frame = CGRect(
                x: x - CGFloat(item.paddingLeft),
                y: 0,
                width: length + CGFloat(item.paddingLeft) + CGFloat(item.paddingRight),
                height: barSize.height
            )
            result.contentBoxes[item.id] = box
        }

        // Left: flows right from the bar's left padding.
        var cursor = CGFloat(settings.paddingLeft)
        for item in items where item.position == .left && item.isVisible && item.isInBarFlow {
            let measured = measure(item)
            cursor += CGFloat(item.paddingLeft)
            place(item, x: cursor, measured: measured)
            cursor += contentLength(item: item, measured: measured) + CGFloat(item.paddingRight)
        }

        // Right: flows left from the bar's right padding.
        cursor = width - CGFloat(settings.paddingRight)
        for item in items where item.position == .right && item.isVisible && item.isInBarFlow {
            let measured = measure(item)
            cursor -= CGFloat(item.paddingRight)
            let length = contentLength(item: item, measured: measured)
            cursor -= length
            place(item, x: cursor, measured: measured)
            cursor -= CGFloat(item.paddingLeft)
        }

        // Center: pre-pass sums the block, then flows right from the centered origin.
        let centerItems = items.filter { $0.position == .center && $0.isVisible && $0.isInBarFlow }
        if !centerItems.isEmpty {
            var measures: [Int: MeasuredContent] = [:]
            var total: CGFloat = 0
            for item in centerItems {
                let measured = measure(item)
                measures[item.id] = measured
                total += CGFloat(item.paddingLeft) + contentLength(item: item, measured: measured)
                    + CGFloat(item.paddingRight)
            }
            cursor = (width - total) / 2
            for item in centerItems {
                let measured = measures[item.id]!
                cursor += CGFloat(item.paddingLeft)
                place(item, x: cursor, measured: measured)
                cursor += contentLength(item: item, measured: measured) + CGFloat(item.paddingRight)
            }
        }

        // centerLeft (q): flows left, starting at the notch's left edge.
        cursor = width / 2 - notchHalf
        for item in items where item.position == .centerLeft && item.isVisible && item.isInBarFlow {
            let measured = measure(item)
            cursor -= CGFloat(item.paddingRight)
            let length = contentLength(item: item, measured: measured)
            cursor -= length
            place(item, x: cursor, measured: measured)
            cursor -= CGFloat(item.paddingLeft)
        }

        // centerRight (e): flows right, starting at the notch's right edge.
        cursor = width / 2 + notchHalf
        for item in items where item.position == .centerRight && item.isVisible && item.isInBarFlow {
            let measured = measure(item)
            cursor += CGFloat(item.paddingLeft)
            place(item, x: cursor, measured: measured)
            cursor += contentLength(item: item, measured: measured) + CGFloat(item.paddingRight)
        }

        // Invisible items get an empty frame so hit testing skips them.
        for item in items where !item.isVisible || !item.isInBarFlow {
            item.frame = .zero
            result.contentBoxes[item.id] = .zero
        }

        return result
    }
}

File: src/model/layout.cpp
#include "model/layout.h"

namespace ybar::model {

double partAdvance(const TextPart& part, const Size& measured) {
    if (!part.drawing) return 0;
    if (part.customWidth >= 0) return part.customWidth;
    return part.paddingLeft + measured.width + part.paddingRight;
}

namespace {

// Component widths BETWEEN icon and label. The image sits outside this run
// (leading, or trailing the label when align=r) but its advance is part of
// the item's length either way.
double componentWidth(const Item& item) {
    double width = 0;
    if (item.image) width += item.image->advance();
    if (item.graph) width += item.graph->capacity; // capacity points, sample_width = 1
    if (item.slider) width += item.slider->width;
    if (item.gauge) width += item.gauge->size;
    return width;
}

} // namespace

double contentLength(const Item& item, const MeasuredContent& measured) {
    if (item.customWidth >= 0) return item.customWidth;
    return naturalLength(item, measured);
}

double naturalLength(const Item& item, const MeasuredContent& measured) {
    const double icon = partAdvance(item.icon, measured.icon);
    // A gauge's label renders centered inside the dial: zero flow width.
    const double label = item.gauge ? 0 : partAdvance(item.label, measured.label);
    return icon + componentWidth(item) + label;
}

std::unordered_map<int, Rect> layout(const std::vector<std::unique_ptr<Item>>& items,
                                     const LayoutSettings& settings, const Measure& measure) {
    std::unordered_map<int, Rect> boxes;
    const double height = settings.height;
    const double notchHalf = settings.notchWidth / 2;

    double leftX = settings.paddingLeft;
    double rightX = settings.width - settings.paddingRight;
    double qX = settings.width / 2 - notchHalf;
    double eX = settings.width / 2 + notchHalf;

    // Center pre-pass: the block is centered as a unit.
    double centerTotal = 0;
    for (const auto& item : items) {
        if (item->position != ItemPosition::Center) continue;
        if (!item->isVisibleInFlow() || item->kind == ItemKind::Bracket) continue;
        const auto m = measure(*item);
        centerTotal += item->paddingLeft + contentLength(*item, m) + item->paddingRight;
    }
    double centerX = (settings.width - centerTotal) / 2;

    for (const auto& item : items) {
        const bool inFlow = item->kind != ItemKind::Bracket &&
                            item->position != ItemPosition::Popup && item->isVisibleInFlow();
        if (!inFlow) {
            item->frame = Rect{};
            boxes[item->id] = Rect{};
            continue;
        }
        const auto m = measure(*item);
        const double len = contentLength(*item, m);
        Rect content;
        switch (item->position) {
            case ItemPosition::Left: {
                leftX += item->paddingLeft;
                content = Rect{leftX, 0, len, height};
                leftX += len + item->paddingRight;
                break;
            }
            case ItemPosition::Right: {
                rightX -= item->paddingRight + len;
                content = Rect{rightX, 0, len, height};
                rightX -= item->paddingLeft;
                break;
            }
            case ItemPosition::Center: {
                centerX += item->paddingLeft;
                content = Rect{centerX, 0, len, height};
                centerX += len + item->paddingRight;
                break;
            }
            case ItemPosition::CenterLeft: { // q: flows LEFT away from the notch
                qX -= item->paddingRight + len;
                content = Rect{qX, 0, len, height};
                qX -= item->paddingLeft;
                break;
            }
            case ItemPosition::CenterRight: { // e: flows RIGHT away from the notch
                eX += item->paddingLeft;
                content = Rect{eX, 0, len, height};
                eX += len + item->paddingRight;
                break;
            }
            case ItemPosition::Popup:
                content = Rect{};
                break;
        }
        boxes[item->id] = content;
        item->frame = Rect{content.x - item->paddingLeft, 0,
                           item->paddingLeft + len + item->paddingRight, height};
    }
    return boxes;
}

Rect bracketFrame(const Item& bracket, const std::vector<Item*>& members,
                  const std::unordered_map<int, Rect>& contentBoxes, double barHeight) {
    Rect box;
    for (const auto* member : members) {
        const auto it = contentBoxes.find(member->id);
        if (it == contentBoxes.end() || it->second.isZero() || it->second.width <= 0) continue;
        box = box.unionWith(it->second);
    }
    if (box.isEmpty()) return Rect{};
    return Rect{box.x - bracket.paddingLeft, 0,
                box.width + bracket.paddingLeft + bracket.paddingRight, barHeight};
}

} // namespace ybar::model
